dcpy/connectors/edm/recipes.py

```diff
@@ -215,11 +261,21 @@ def get_source_data_versions(recipe: Recipe):
     ]
 
 
+def _apply_recipe_defaults(recipe: Recipe):
+    recipe.inputs.dataset_defaults = recipe.inputs.dataset_defaults or DatasetDefaults()
+
+    for ds in recipe.inputs.datasets:
+        ds.preprocessor = ds.preprocessor or recipe.inputs.dataset_defaults.preprocessor
+        ds.file_type = ds.file_type or recipe.inputs.dataset_defaults.file_type
+
+
 def recipe_from_yaml(path: Path) -> Recipe:
     """Import a recipe file from yaml, and validate schema."""
     with open(path, "r", encoding="utf-8") as f:
         s = yaml.safe_load(f)
-    return Recipe(**s)
+        recipe = Recipe(**s)
+        _apply_recipe_defaults(recipe)
```

Comment:
Yeah, I definitely prefer not mutating in general. Python doesn't make that as easy as say, F# or Clojure, at least with the class-based approach. But this has me pondering better ways to deserialize these files... Would prefer not to even have to set defaults.

In this case, since it was cumbersome to actually write this with an immutable-data approach, I wanted to make it obvious that the argument to `_apply_recipe_defaults` is being mutated, which is less obvious if we're returning a value from that function.

Reply:
Ahh, that makes sense. I didn't quite think about the mechanics of returning a new object with a single changed field. Sounds good!

Reply:
Not to be a broken record, but fsharp is great here. If you have
```fsharp
let myObj = {
    field1 = "string"
    field2 = "otherstring"
    field3 = 1
)
```
You can just
```fsharp
let myUpdatedObj1 = { myObj with field1 = "newString" }
let myUpdatedObj2 = { 
    myObj with 
        field2 = "updatedstring"
        field3 = 2
}
```